lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -196,6 +219,36 @@ newStakePoolLayer nl db@DBLayer {..} =
     _knownPools =
         Set.fromList <$> liftIO (atomically listRegisteredPools)
 
+    -- In order to apply the settings, we have to restart the
+    -- metadata sync thread as well to avoid race conditions.
+    _putSettings :: TVar ThreadId -> Settings -> IO ()
+    _putSettings tv settings = do
+        -- stop syncing thread, so we can apply the settings cleanly
+        mt <- readTVarIO tv
+        killThread mt
```

Comment:
That thread has no particular exception handler and who knows what will happens if we kill it in the middle of an HTTP query or in the middle of a database call. May I suggest to not kill the thread and to revert the control here? What about:

- Passing an `MVar Settings` to the fetching thread. 
- Have the fetching thread taking that MVar before every fetch, and putting it back once fetched and done (before going to sleep, using a safe bracket-style like `withMVar`). 
- Have the `putSettings` function also take and replace that MVar when handling a query, replacing the content with the new settings, and dropping the table if necessary while holding on that MVar. 

This way, we get rid of potential errors that will occur from killing the thread at arbitrary time, and we still ensure complete synchronization of both thread?

Reply:
> That thread has no particular exception handler and who knows what will happens if we kill it in the middle of an HTTP query or in the middle of a database call. 

What exactly can go wrong here? You'd have to be concerned about every ctrl-c by a user as well then, because it causes an async exception too. I'd exepect database operations to be atomic.

Reply:
> You'd have to be concerned about every ctrl-c by a user as well the user

We are! Precisely! I think we should avoid killing thread were we don't need to. Async exceptions are hard to deal with as they are without us shooting ourselves in the foot. Killing the thread arbitrarily may result in random errors (which won't probably do anything more than printing an error log, but that's precisely what I want to avoid). We shouldn't raise errors when there are easy ways to avoid them.

Reply:
Ok can. I think we need to be careful in the endpoint tests then, because of the potential delay.

Reply:
I'm not really liking dealing with MVars directly. It's very easy to break and end up with deadlocks, infinite loops etc etc